### PR TITLE
FEAT: Support tcp store for rendezvous in collective communication

### PR DESCRIPTION
**Pull request by ChengjieLi28**
Related #22 

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.11...3.21)
+
+project(XoscarCollective)
+
+if(NOT ${PYTHON_EXECUTABLE})
```
Comment (codingl2k1) [inline CMakeLists.txt:5]: The `PYTHON_EXECUTABLE` may not be found, then the Python executable will be incorrect.

Comment (ChengjieLi28) [reply CMakeLists.txt]: Fix

File: cpp/collective/rendezvous/src/socket.cpp [hunk 2]
```diff
@@ -0,0 +1,919 @@
...
+const std::chrono::seconds SocketConnectOp::delay_duration_{1};
+
+SocketConnectOp::SocketConnectOp(const std::string &host,
+                                 std::uint16_t port,
+                                 const SocketOptions &opts)
+    : host_{host.c_str()}, port_{fmt::to_string(port)}, opts_{&opts} {}
+
+std::unique_ptr<SocketImpl> SocketConnectOp::run() {
+    if (opts_->prefer_ipv6()) {
+        //    xoscar_DEBUG(
```
[the remark this answers is not in the document]
Comment (ChengjieLi28) [reply cpp/collective/rendezvous/src/socket.cpp:608]: Here's an issue about logging module #35 . Do it later.

File: cpp/collective/rendezvous/src/tcp_store.cpp [hunk 3]
```diff
@@ -0,0 +1,1312 @@
...
+    client_ = detail::TCPClient::connect(addr_, opts);
+
+    if (opts.waitWorkers) {
+        waitForWorkers();
+    }
+
+    callbackClient_ = detail::TCPCallbackClient::connect(addr_, opts);
+}
+
+TCPStore::~TCPStore() = default;
```
[the remark this answers is not in the document]
Comment (ChengjieLi28) [reply cpp/collective/rendezvous/src/tcp_store.cpp:1103]: Add a comment here. In ``Torch``, there is no recycle for the ``TCPStore`` object for the interface ``init_process_group``. See ``_create_c10d_store`` function of ``rendezvous.py`` in ``PyTorch``.
So just remain the origin implement.

**Pull request merged**